doc/designs/quic-design/quic-connection-id-cache.md

```diff
@@ -0,0 +1,204 @@
+
+/*
+ * Allocate and free a connection ID cache
+ */
+QUIC_CONN_ID_CACHE *ossl_quic_new_conn_id_cache(void);
+void ossl_quic_free_conn_id_cache(QUIC_CONN_ID_CACHE *cache);
+
+/*
+ * Add an ID to connection mapping to the cache.
+ * One call is to add our local ID, the other to add our peer's remote ID.
+ */
+int ossl_quic_set1_local_conn_id(QUIC_CONN_ID_CACHE *cache,
+                                 const SSL_CONNECTION *conn,
+                                 const unsigned char *conn_id,
+                                 size_t conn_id_len);
+int ossl_quic_set1_remote_conn_id(QUIC_CONN_ID_CACHE *cache,
+                                  const SSL_CONNECTION *conn,
+                                  const unsigned char *conn_id,
+                                  size_t conn_id_len);
+
+/*
+ * Remove an ID to connection mapping to the cache.
+ * One call is to remove our local ID, the other to remove our peer's remote ID.
+ */
+int ossl_quic_remove_local_conn_id(QUIC_CONN_ID_CACHE *cache,
+                                   const unsigned char *conn_id,
+                                   size_t conn_id_len);
+int ossl_quic_remove_remote_conn_id(QUIC_CONN_ID_CACHE *cache,
+                                    const unsigned char *conn_id,
+                                    size_t conn_id_len);
+
+/*
+ * Remove a connection and all associated cached IDs
+ */
+int ossl_quic_remove_connection(QUIC_CONN_ID_CACHE *cache,
+                                const SSL_CONNECTION *conn);
+
+/*
+ * Lookup a connection by ID.
+ * One call is to lookup based on a local ID, the other to lookup based our
+ * peer's remote ID.
+ */
+SSL_CONNECTION *ossl_quic_get0_local_conn_id(const QUIC_CONN_ID_CACHE *cache,
+                                             const unsigned char *conn_id,
+                                             size_t conn_id_len);
+SSL_CONNECTION *ossl_quic_get0_remote_conn_id(const QUIC_CONN_ID_CACHE *cache,
+                                              const unsigned char *conn_id,
+                                              size_t conn_id_len);
+
+/*
+ * Query the most recent ID for a connection.
+ * One call is to query the most recent local ID, the other to query the most
+ * recent remote ID.
+ *
+ * These would better be stored in the connection itself.
+ */
+int ossl_get0_local_conn_id(const QUIC_CONN_ID_CACHE *cache,
+                            const SSL_CONNECTION *conn,
+                            unsigned char **conn_id,
+                            size_t *conn_id_len);
+int ossl_get0_remote_conn_id(const QUIC_CONN_ID_CACHE *cache,
+                             const SSL_CONNECTION *conn,
+                             unsigned char **conn_id,
+                             size_t *conn_id_len);
+```
+
+MVP
+---
+
+For MVP the cache only needs one connection ID and the function implementations
+need only be minimal.  Something along the lines of the follow pseudo code
+would be enough.
```

Comment:
Actually no, see the third paragraph in 5.1:

```
Multiple connection IDs are used so that endpoints can send packets
that cannot be identified by an observer as being for the same connection
without cooperation from an endpoint;
see [Section 9.5](https://www.ietf.org/rfc/rfc9000.html#migration-linkability).
```

Reply:
Hmmmm, this greatly complicates things.  Doing this will require effectively a complete implementation.  I'm fine doing that but was hoping to reduce the immediate required effort.

I don't think that the sequence number and retirement should be part of the connection ID cache.  The cache is mapping connection ID to SSL conection objects.  I'd have thought it was the connection object that should track this.  That said, having these as part of the cache isn't difficult.  It just seems out of place for the module.

Reply:
The sequence number is 1 to 1 mapped to a connection ID. I do not know where else it should be stored other than in the database of known connection IDs which is effectively the connection ID cache. Of course there is also the _latest sequence number_ and _latest retired sequence number_ which probably should be stored in the connection object.